### PR TITLE
S4057: Add support for tuples and assignment and declaration in the same deconstruction

### PR DESCRIPTION
**Pull request by csaba-sagi-sonarsource**
Please do not review `MapAssignmentArguments` It will be replaced by the method from work on rule S4055.

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs [hunk 1]
```diff
@@ -118,15 +135,105 @@ public sealed class SetLocaleForDataTypes : SonarDiagnosticAnalyzer
             {
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs:135]: Unrelated to the PR:
The line above `objectCreation.GetFirstNonParenthesizedParent()` seems to be equivalent to `WalkUpParentheses` I created recently. Can you remove the "WalkUpParentheses" method and replace the invocation with "GetFirstNonParenthesizedParent". Thanks.

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
Can be simplified at some points. Besides, that looks good. The argument to target mapping should be made an extension method, as it is useful in other contexts too.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs [hunk 2]
```diff
@@ -118,15 +135,105 @@ public sealed class SetLocaleForDataTypes : SonarDiagnosticAnalyzer
...
+                    {
+                        currentNode = currentNode.Parent;
+                    }
+                }
+                return FindAssignedSyntaxNodeInAssignment(argument, assignment);
+            }
+            return null;
+        }
+
+        private static SyntaxNode FindAssignedSyntaxNodeInAssignment(ArgumentSyntax argument, AssignmentExpressionSyntax assignment)
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs:165]: Consider using [IDeconstructionAssignmentOperation](https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.operations.ideconstructionassignmentoperation?view=roslyn-dotnet-3.2.1) for this. I didn't look into the details, but it has `Target` and `Value`. I also don't know how good the IOperation part of the shim layer is.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SetLocaleForDataTypes.CSharp10.cs [hunk 3]
```diff
@@ -2,6 +2,28 @@
...
 using System.Globalization;
 
-DataTable y;
+DataTable y, d, f, j, l;
+int b, h;
 
-(var x, y) = (new DataTable(), new DataTable()); // FN
+(var x, y) = (new DataTable(), new DataTable()); // Noncompliant
+// Noncompliant@-1
+
```
[the remark this answers is not in the document]
Comment (csaba-sagi-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SetLocaleForDataTypes.CSharp10.cs:10]: I do not understand your suggestion, could you please describe what was the intention here?

File: analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs [hunk 4]
```diff
@@ -83,25 +84,41 @@ public sealed class SetLocaleForDataTypes : SonarDiagnosticAnalyzer
...
+            };
 
-                if (GetPropertySymbol(assignmentExpression, c.SemanticModel) is { } propertySymbol
-                    && propertySymbol.Name == "Locale"
-                    && propertySymbol.ContainingType.IsAny(CheckedTypes))
+        private static void ProcessExpression(ExpressionSyntax expression, SemanticModel semanticModel, ISet<ISymbol> symbolsWhereLocaleIsSet)
+        {
+            expression = expression.RemoveParentheses();
+            if (TupleExpressionSyntaxWrapper.IsInstance(expression)
+                && ((TupleExpressionSyntaxWrapper)expression) is var tupleExpression)
```
[the remark this answers is not in the document]
Comment (csaba-sagi-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs:94]: In theory we can change it, but it will be a pain to update it in the future, so suggest not to change it. Only minor things like adding values to SyntaxKindEx for example.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs [hunk 5]
```diff
@@ -70,6 +70,7 @@ public sealed class SetLocaleForDataTypes : SonarDiagnosticAnalyzer
                     && GetAssignmentTargetVariable(c.Node) is { } variableSyntax)
```
[the remark this answers is not in the document]
Comment (csaba-sagi-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs:70]: Once all the helper methods get merged, it should be checked what can be used from there.

**Comment by csaba-sagi-sonarsource** [comment]
Please @mary-georgiou-sonarsource  also add a test case like `var (x,y) = (new DataTable(), new DataTable()); ` It should work out of the box once we start using the extension method from the S4055 PR.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SetLocaleForDataTypes.CSharp10.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SetLocaleForDataTypes.CSharp10.cs:10]: I meant to add another test case with `var (...=)` besides `(var x, var y)`

File: analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs:94]: @costin-zaharia-sonarsource I was talking with Pavel in our onboarding meeting about this and he was suggesting to ask you about this. Can we change the code-generation logic or is the generated code copy/paste from Tunnel Vision. Especially can we run the script `eng/generate-compiler-code.cmd` mentioned here and change the output from it?

 https://github.com/SonarSource/sonar-dotnet/blob/b556a308bcf88c4bdefa8c47805215f9a58d4b75/analyzers/src/SonarAnalyzer.CFG/ShimLayer/Syntax.xml#L1-L11

[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs:94]: @costin-zaharia-sonarsource and @pavel-mikula-sonarsource I found the place, where `IsInstance` is generated:

https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/9c5d064f100186515bcb45aadd629f18bc523a28/StyleCop.Analyzers/StyleCop.Analyzers.CodeGeneration/SyntaxLightupGenerator.cs#L707-L743

sharwell uses a code generator and generates the code based on `Syntax.xml` (see [here](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/9c5d064f100186515bcb45aadd629f18bc523a28/StyleCop.Analyzers/StyleCop.Analyzers.CodeGeneration/SyntaxLightupGenerator.cs#L86-L97)). I don't know whether it is worth integrating the whole generator into our project somehow.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SetLocaleForDataTypes.CSharp10.cs [hunk 6]
```diff
@@ -2,6 +2,28 @@
...
+(var g, h) = (new DataTable(), 42);
+(g.Locale, h) = (CultureInfo.InvariantCulture, 42);
+
+(var i, j) = (42, new DataTable());
+(i, j.Locale) = (42, CultureInfo.InvariantCulture);
+
+(var k, l) = (new DataTable(), new DataTable()); // Noncompliant
+k.Locale = CultureInfo.InvariantCulture;
+
+TupleParameter((new DataTable(), new DataTable()));
```
[the remark this answers is not in the document]
Comment (mary-georgiou-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SetLocaleForDataTypes.CSharp10.cs]: Added and opened trello cards. I'll open issues for all the unsolved ones at the end of the sprint.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SetLocaleForDataTypes.CSharp10.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mary-georgiou-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SetLocaleForDataTypes.CSharp10.cs:9]: added

File: analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs [hunk 7]
```diff
@@ -83,25 +84,41 @@ public sealed class SetLocaleForDataTypes : SonarDiagnosticAnalyzer
...
                 var assignmentExpression = (AssignmentExpressionSyntax)c.Node;
+                ProcessExpression(assignmentExpression.Left, c.SemanticModel, symbolsWhereLocaleIsSet);
+            };
 
-                if (GetPropertySymbol(assignmentExpression, c.SemanticModel) is { } propertySymbol
-                    && propertySymbol.Name == "Locale"
-                    && propertySymbol.ContainingType.IsAny(CheckedTypes))
+        private static void ProcessExpression(ExpressionSyntax expression, SemanticModel semanticModel, ISet<ISymbol> symbolsWhereLocaleIsSet)
+        {
+            expression = expression.RemoveParentheses();
```
[the remark this answers is not in the document]
Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs]: So this is an FP for now. But I'm wondering - do we really need to support this case?

File: analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs (hunk 2 above)
Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs]: This method  does not exist anymore.

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs (hunk 1 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs:123]: I created #5790

File: analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs [hunk 8]
```diff
@@ -113,20 +118,20 @@ public sealed class SetLocaleForDataTypes : SonarDiagnosticAnalyzer
...
-        private static SyntaxNode GetAssignmentTargetVariable(SyntaxNode objectCreation) =>
+        private static ImmutableArray<SyntaxNode> GetAssignmentTargetVariable(SyntaxNode objectCreation) =>
             objectCreation.GetFirstNonParenthesizedParent() switch
             {
-                AssignmentExpressionSyntax assignment => assignment.Left,
-                EqualsValueClauseSyntax equalsClause when equalsClause.Parent.Parent is VariableDeclarationSyntax variableDeclaration => variableDeclaration.Variables.Last(),
-                _ => null
+                AssignmentExpressionSyntax assignment => assignment.AssignmentTargets(),
+                EqualsValueClauseSyntax equalsClause when equalsClause.Parent.Parent is VariableDeclarationSyntax variableDeclaration =>
+                    ImmutableArray.Create((SyntaxNode)variableDeclaration.Variables.Last()),
```
[the remark this answers is not in the document]
Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs]: Both issues are raised for the correct object creation - but the second one is not raised in the correct location.

File: analyzers/src/SonarAnalyzer.Common/Common/NodeAndSymbol.cs [hunk 9]
```diff
@@ -34,7 +34,12 @@ public NodeAndSymbol(TSyntaxNode node, TSymbol symbol)
             Node = node;
             Symbol = symbol;
         }
-    }
+
+        public override bool Equals(object obj) =>
+            obj is NodeAndSymbol other && Symbol.Equals(other.Symbol);
```
[the remark this answers is not in the document]
Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Common/NodeAndSymbol.cs:39]: I removed the changes from NodeAndSymbol class.
Indeed it's more clear to me as well with a Dictionary.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs [hunk 10]
```diff
@@ -113,20 +112,20 @@ public sealed class SetLocaleForDataTypes : SonarDiagnosticAnalyzer
...
-        private static SyntaxNode GetAssignmentTargetVariable(SyntaxNode objectCreation) =>
+        private static ImmutableArray<SyntaxNode> GetAssignmentTargetVariables(SyntaxNode objectCreation) =>
             objectCreation.GetFirstNonParenthesizedParent() switch
             {
-                AssignmentExpressionSyntax assignment => assignment.Left,
-                EqualsValueClauseSyntax equalsClause when equalsClause.Parent.Parent is VariableDeclarationSyntax variableDeclaration => variableDeclaration.Variables.Last(),
-                _ => null
+                AssignmentExpressionSyntax assignment => assignment.AssignmentTargets(),
+                EqualsValueClauseSyntax equalsClause when equalsClause.Parent.Parent is VariableDeclarationSyntax variableDeclaration =>
+                    variableDeclaration.Variables.Select(x => (SyntaxNode)x).ToImmutableArray(),
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs:120]: I was wrong here: `variableDeclaration.Variables.Last()` is actually right. The "=" only assigns to the last variable. See e.g. https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAIwCwAUPgMwAERNAwjQN4U3t2GEtsd8D0/WAhpQaAXhrE4NAFYSahANy8+7YaJnzJy1WvzEAnAAoAJACIo45lAC+W67NvmAlCvJ9bFW0A==

File: analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs [hunk 11]
```diff
@@ -113,20 +112,20 @@ public sealed class SetLocaleForDataTypes : SonarDiagnosticAnalyzer
...
                 : semanticModel.GetSymbolInfo(constructorCall).Symbol?.ContainingType;
 
-        private static SyntaxNode GetAssignmentTargetVariable(SyntaxNode objectCreation) =>
+        private static ImmutableArray<SyntaxNode> GetAssignmentTargetVariables(SyntaxNode objectCreation) =>
             objectCreation.GetFirstNonParenthesizedParent() switch
             {
-                AssignmentExpressionSyntax assignment => assignment.Left,
-                EqualsValueClauseSyntax equalsClause when equalsClause.Parent.Parent is VariableDeclarationSyntax variableDeclaration => variableDeclaration.Variables.Last(),
-                _ => null
+                AssignmentExpressionSyntax assignment => assignment.AssignmentTargets(),
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs:118]: The direct parent of the `new` is an assignment. In this case `assignment.Left` is the right choice

File: analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs [hunk 12]
```diff
@@ -113,20 +112,20 @@ public sealed class SetLocaleForDataTypes : SonarDiagnosticAnalyzer
...
+                EqualsValueClauseSyntax equalsClause when equalsClause.Parent.Parent is VariableDeclarationSyntax variableDeclaration =>
+                    variableDeclaration.Variables.Select(x => (SyntaxNode)x).ToImmutableArray(),
+                ArgumentSyntax argument when argument.Ancestors().OfType<AssignmentExpressionSyntax>().FirstOrDefault() is { } expressionSyntax =>
+                    expressionSyntax.AssignmentTargets(),
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs:122]: A perfect use case for this:
https://github.com/SonarSource/sonar-dotnet/blob/9f42caa9d16c9c2b8867d05a17809843f9477800/analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs#L133-L141

It is part of #5798

File: analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs [hunk 13]
```diff
@@ -113,20 +112,20 @@ public sealed class SetLocaleForDataTypes : SonarDiagnosticAnalyzer
...
+                _ => ImmutableArray<SyntaxNode>.Empty
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs:123]: I think you should use `FindAssignmentComplement` here. With that change, `private static ImmutableArray<SyntaxNode> GetAssignmentTargetVariables(SyntaxNode objectCreation)` can be changed backed to return a single SyntaxNode. There are some obscure patterns, that allow you to can assign to more than one variable, but these are out of scope of the PR. e.g.

```cs
        object i, j;
        i = (j = new object()); // i and j point to the same value now
        var k = (j ??= new object()); // This is a bit more popular recently for caching.
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs [hunk 14]
```diff
@@ -113,20 +112,20 @@ public sealed class SetLocaleForDataTypes : SonarDiagnosticAnalyzer
...
             };
 
-        private static IPropertySymbol GetPropertySymbol(AssignmentExpressionSyntax assignment, SemanticModel model) =>
-            model.GetSymbolInfo(assignment.Left).Symbol as IPropertySymbol;
-
-        private static ISymbol GetAccessedVariable(AssignmentExpressionSyntax assignment, SemanticModel model)
+        private static ISymbol GetAccessedVariable(SyntaxNode node, SemanticModel model)
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs:126]: This method should use the same logic as GetAssignmentTargetVariables; let's review it once GetAssignmentTargetVariables is in a better state.

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SetLocaleForDataTypes.CSharp10.cs [hunk 15]
```diff
@@ -2,6 +2,28 @@
...
+(var c, d) = (42, new DataTable()); // Noncompliant
+
+(var e, f) = (new DataTable(), new DataTable());
+(e.Locale, f.Locale) = (CultureInfo.InvariantCulture, CultureInfo.InvariantCulture);
+
+(var g, h) = (new DataTable(), 42);
+(g.Locale, h) = (CultureInfo.InvariantCulture, 42);
+
+(var i, j) = (42, new DataTable());
+(i, j.Locale) = (42, CultureInfo.InvariantCulture);
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SetLocaleForDataTypes.CSharp10.cs:21]: I would still add one test.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs]: Created #5798

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SetLocaleForDataTypes.CSharp10.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (mary-georgiou-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SetLocaleForDataTypes.CSharp10.cs:21]: I'm not sure what was meant here exactly - but I added some UTs with nested tuples.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs]: @andrei-epure-sonarsource

File: analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs [hunk 16]
```diff
@@ -62,19 +61,25 @@ public sealed class SetLocaleForDataTypes : SonarDiagnosticAnalyzer
...
                 });
 
-        private static Action<SyntaxNodeAnalysisContext> ProcessObjectCreations(ISet<NodeAndSymbol> symbolsWhereTypeIsCreated) =>
+        private static Action<SyntaxNodeAnalysisContext> ProcessObjectCreations(IDictionary<ISymbol, SyntaxNode> symbolsWhereTypeIsCreated) =>
             c =>
             {
                 if (GetSymbolFromConstructorInvocation(c.Node, c.SemanticModel) is ITypeSymbol objectType
                     && objectType.IsAny(CheckedTypes)
-                    && GetAssignmentTargetVariable(c.Node) is { } variableSyntax)
+                    && GetAssignmentTargetVariables(c.Node) is { } variableSyntax)
```
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs:69]: why is `GetAssignmentTargetVariables` plural if there's only a single result

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SetLocaleForDataTypes.CSharp10.cs [hunk 17]
```diff
@@ -2,6 +2,40 @@
...
+
+var dt = new DataTable();
+(i, (dt.Locale, h)) = (42, (CultureInfo.InvariantCulture, 0));
+
+(var k, l) = (new DataTable(), new DataTable()); // Noncompliant
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SetLocaleForDataTypes.CSharp10.cs:26]: here the location would be useful to ensure the right one is raised

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SetLocaleForDataTypes.CSharp10.cs [hunk 18]
```diff
@@ -2,6 +2,40 @@
...
+k.Locale = CultureInfo.InvariantCulture;
+
+var (xx, yy) = (new DataTable(), new DataTable()); // Noncompliant [3, 4]
+var (xxx, yyy) = (0, (new DataTable(), 2)); // FN
+(int xxxx, (DataTable, int) yyyy) = (0, (new DataTable(), 2)); // FN
+
+TupleParameter((new DataTable(), new DataTable())); // FN
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SetLocaleForDataTypes.CSharp10.cs:33]:
```suggestion
TupleParameter((new DataTable(), new DataTable())); // passed as parameter to a method
```

**Comment by andrei-epure-sonarsource** [comment]
@csaba-sagi-sonarsource I am missing quite a lot of the context, could you please have another look.

And I guess @martin-strecker-sonarsource can have a look at it on Monday

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
There are potentially some problems with `GetAccessedVariable`. Add some tests to see which cases are broken (if any).

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SetLocaleForDataTypes.CSharp10.cs [hunk 19]
```diff
@@ -40,3 +40,8 @@
 DataTable foo, bar;
 foo = (bar = new DataTable()); // Noncompliant
 var foobar = (bar ??= new DataTable()); // FN
+
+DataTable myTable5, myTable6;
+var (myTable1, myTable2) = ((new DataTable { Locale = CultureInfo.InvariantCulture }), new DataTable()); // Noncompliant
+(var  myTable3, var  myTable4) = (new DataTable { Locale = CultureInfo.InvariantCulture }, new DataTable()); // Noncompliant
+(myTable5, myTable6) = (new DataTable { Locale = CultureInfo.InvariantCulture }, new DataTable()); // Noncompliant
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SetLocaleForDataTypes.CSharp10.cs:47]: Please add ^ markers for the diagnostic positions.

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs]: Can we do something about this? Should be fairly simple.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs [hunk 20]
```diff
@@ -117,38 +118,44 @@ public sealed class SetLocaleForDataTypes : SonarDiagnosticAnalyzer
...
-
-                var lastVariable = identifier.FirstAncestorOrSelf<VariableDeclarationSyntax>()?.Variables.LastOrDefault();
-                return lastVariable != null
-                    ? model.GetDeclaredSymbol(lastVariable)
-                    : null;
+                leftSideOfParentAssignment = argumentSyntax.FindAssignmentComplement();
+            }
+            else
+            {
+                leftSideOfParentAssignment = objectCreation.FirstAncestorOrSelf<AssignmentExpressionSyntax>()?.Left;
```
[the remark this answers is not in the document]
Comment (csaba-sagi-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs:146]: The UT was failing because I was missing some code in the if branch. so it cannot be written that nice anymore.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs [hunk 21]
```diff
@@ -83,27 +88,23 @@ public sealed class SetLocaleForDataTypes : SonarDiagnosticAnalyzer
...
+                    .Where(x => c.SemanticModel.GetSymbolInfo(x).Symbol is IPropertySymbol propertySymbol
+                                && propertySymbol.Name == "Locale"
+                                && propertySymbol.ContainingType.IsAny(CheckedTypes))
+                    .Select(x => GetAccessedVariable(x, c.SemanticModel))
+                    .WhereNotNull();
+                symbolsWhereLocaleIsSet.UnionWith(variableSymbols);
             };
 
-        private static Action<CompilationAnalysisContext> ProcessCollectedSymbols(ICollection<NodeAndSymbol> symbolsWhereTypeIsCreated, ICollection<ISymbol> symbolsWhereLocaleIsSet) =>
+        private static Action<CompilationAnalysisContext> ProcessCollectedSymbols(IDictionary<ISymbol, SyntaxNode> symbolsWhereTypeIsCreated, ICollection<ISymbol> symbolsWhereLocaleIsSet) =>
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs:100]:
```suggestion
        private static Action<CompilationAnalysisContext> ProcessCollectedSymbols(IDictionary<ISymbol, SyntaxNode> symbolsWhereTypeIsCreated, ISet<ISymbol> symbolsWhereLocaleIsSet) =>
```

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]

**Review by martin-strecker-sonarsource — APPROVED** [review]
Please have a look at this line from the code coverage report:
* line 103: It is worrisome, that `!symbolsWhereLocaleIsSet.Contains(x.Key)` is only ever true or false. I don't think this is the case. Can you make sure, this is sometimes true and sometimes false by placing a breakpoint here?

**Comment by csaba-sagi-sonarsource** [comment]
> Please have a look at this line from the code coverage report:
> 
> * line 103: It is worrisome, that `!symbolsWhereLocaleIsSet.Contains(x.Key)` is only ever true or false. I don't think this is the case. Can you make sure, this is sometimes true and sometimes false by placing a breakpoint here?

It is the coverage tool that is messing up something as it definitely has both true and false values.
